### PR TITLE
`General`: Allow instructors to import and export grading keys with csv files

### PR DESCRIPTION
**Pull request by ole-ve**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

This pull request is linked to Issue #3998. As changing the grade steps on the web app is cumbersome for the user, the motivation is to make adjusting the grading schema more user-friendly.

### Description

This feature allows the user to import grade steps from a CSV file. Grade Type will be set automatically after importing, depending on the used CSV file structure. In the case of Grade Type is 'Grade', the first passing grade will also be evaluated.
In addition, the user can export the grade steps to a CSV file which will be downloaded.
For both functionalities, there is a button for the user on the UI.

Note: The values for setting the inclusivity as well as points are predefined after importing but can be set afterward. This decision has been made as this seems like the user-friendlier solution than setting these values for every single grade step in the CSV file.

### File Structure

1. Grade Type is Bonus
`bonusPoints,lowerBoundPercentage,upperBoundPercentage,isPassingGrade`

2. Grade Type is Grade
`bonusPoints,lowerBoundPercentage,upperBoundPercentage`

- gradeName: string
- bonusPoints: number
- lowerBoundPercentage: number
- upperBoundPercentage: number
- isPassingGrade: boolean

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration
3. Click Button 'Grading'
4. Generate Default Grade Steps / create default grade steps
5. Export to CSV
6. Refresh again and import the exported file

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
grading-system.component.ts: 95%

### Screenshots
image

### CODE REVIEW

**Review by FlorianGlombik — DISMISSED** [review]
The error message does not really fit when entering a non valid value for upperBoundPoints (negative numbers). 
image

**Review by Hialus — APPROVED** [review]
Tested on TS1. Seems to work fine.
Well done 👍🏻

File: src/main/webapp/app/grading-system/grading-system.component.ts [hunk 1]
```diff
@@ -613,4 +631,137 @@ export class GradingSystemComponent implements OnInit {
...
+            (csvGradeStep) =>
+                ({
+                    gradeName:
+                        gradeType === GradeType.GRADE
+                            ? csvGradeStep[csvColumnsGrade.gradeName]
+                                ? String(csvGradeStep[csvColumnsGrade.gradeName])
+                                : ''
+                            : csvGradeStep[csvColumnsBonus.bonusPoints]
+                            ? String(csvGradeStep[csvColumnsBonus.bonusPoints])
+                            : '',
```
Comment (b-fein) [inline src/main/webapp/app/grading-system/grading-system.component.ts:697]: The triple ternary operator makes this very hard to read and understand what is going on.

**Review by b-fein — COMMENTED** [review]
Code Review: Only found one issue where the code’s readability could be improved.

Tests use async instead of the usual fakeAsync, not sure if that is okay.

**Review by MDK5 — APPROVED** [review]
Tested in testing session on ts1, worked well. Just two minor suggestions:
grafik
These buttons are very different in style, you could try to make it look at bit more unified

When uploading a large file, the program attempts to create a lot of rows in the table. Some check to prevent the site from crashing if someone uploads the wrong file would be helpful

Both are not essential, though

**Review by daniels98it — APPROVED** [review]
Except from previously mentioned suggestions, everything worked fine 👍

File: src/main/webapp/app/grading-system/grading-system.component.ts (hunk 1 above)
Comment (ole-ve) [reply src/main/webapp/app/grading-system/grading-system.component.ts]: Resolved in commit 588e8c1.

**Review by Hialus — APPROVED** [review]
Retested on TS1.
Still works as it should.
The only suggestion I would have is to display an error message when the file was too long to be imported. But since this feature is mostly to prevent the website from crashing when importing the wrong file this isn't necessary.

**Review by MDK5 — APPROVED** [review]
Re-Tested on ts1, still worked fine and both suggestions got implemented :+1:

**Review by MDK5 — APPROVED** [review]
Re-Approve without tests since only tests have been changed. The test changes look good to me.

**Review by Hialus — APPROVED** [review]
Reapprove since only the tests changed.

**Review by cremertim — DISMISSED** [review]
tested on ts 1. everything worked as expected according to the testing steps . clgtm

File: src/main/webapp/app/grading-system/grading-system.component.html [hunk 2]
```diff
@@ -12,6 +12,17 @@ <h2 style="margin-bottom: 0.5em">
...
+            <label class="btn btn-primary me-1 mb-1">
+                <span jhiTranslate="artemisApp.gradingSystem.csv.importButton">Import (.csv)</span>
+                <input type="file" accept=".json,.csv" style="display: none" (change)="onCSVFileSelect($event)" />
+            </label>
+        </div>
+        <div class="d-flex">
+            <button (click)="exportGradingStepsToCsv()" class="btn btn-primary me-1 mb-1" type="button">
+                <span jhiTranslate="artemisApp.gradingSystem.csv.exportButton">Export (.csv)</span>
+            </button>
+        </div>
```
Comment (jpbernius) [inline src/main/webapp/app/grading-system/grading-system.component.html:25]: **Layout:** Maybe put all buttons in one line?

Comment (ole-ve) [reply src/main/webapp/app/grading-system/grading-system.component.html:25]: image

**Review by Hialus — APPROVED** [review]
Reapprove since only the button position changed.

**Review by willergeorg — APPROVED** [review]
Button alignment lgtm

**Review by MDK5 — APPROVED** [review]
Re-Approve, the button positions look good on ts1

**Pull request merged**